bson/unmarshal_test.go

```diff
@@ -429,3 +429,34 @@ func TestUnmarshalBSONWithUndefinedField(t *testing.T) {
 		})
 	}
 }
+
+type BSONDocumentArray struct {
+	Array []D `bson:"array"`
+}
+
+type BSONArray struct {
+	Array bsoncore.Array `bson:"array"`
+}
+
+func TestBsoncoreArrayMarshal(t *testing.T) {
```

Comment:
Given that this doesn't use `Unmarshal`, it should go in `marshal_test.go`, but I also think it would instead be better to write a test that uses both and add it to `bson_test.go`. Perhaps some sort of round-trip test that starts with a hand-build bsoncore array and then calls both `Marshal` and `Unmarshal` on it.

You should also add tests to `default_value_encoders_test.go` and `default_value_decoders_test.go` for the new codec.

Reply:
I think this test might already be the one you're describing. It uses both `Marshal` and `Unmarshal` to do conversions between `[]bson.D`, regular BSON, and `bsoncore.Array`, and asserts that there's been no change to the underlying BSON. It's just a simplification of the context given in the original ticket.

Also, added tests to `default_value_encoders_test` and `default_value_decoders_test`.